### ISSUE TITLE
feedback/issues Zendesk import

### ISSUE THREAD

**Issue by qschneider** [zammad/zammad#558]
Hi @thorsteneckel this is the reply to your changes in #487.

Great job, thanks for the effort ! 

I tested this on a fresh pulled docker image.

Import went pretty well and after import i was able to login with my credentials as described in your comment. Also the import quality is much better, no more HTML-tags inside the ticket thread.

From what i can tell, there are still some minor bugs, that may need your attention.

- imho - before first login (and after migration) there should be again the information regarding the first time credentials, otherwise someone who overlocked it during migration, might be confused. Or maybe place the notice only during first time login, because thats the time someone will need that information anyway.

- you have to place a trailing / in the url, otherwise my api token was not accepted by the migration wizard. Maybe you could include this in your url-check and place the trailing / if not provided by the user.

- migration speed still feels slow (10 Minutes for 40 users and 80 tickets). being so slow, there should be any kind of indicator that is showing the user that the import-process is still running, otherwise users might refresh the page or just close it - not sure what will happen with the import in that case.

- archived tickets aren't migrated - not sure at what point zendesk flags a ticket as archived. It is still accessible in the zendesk client by viewing a persons ticket history, but not in the regular agent view. If this is an api-restriction, users should be informed on that limitation before starting the migration.

- The owner isn't set in any of the imported tickets, is this done by purpose ? It just feels uncommon for someone who is transitioning from zendesk , because the owner or agent setting is mandatory there.

There are some more things that feels different after migration (or not as productive as it feels in zendesk) which i will address in another ticket.

Anyway, thanks for the great work to you and the whole team.

**Comment by thorsteneckel** [comment]
Hi @qschneider - wow! Thanks for the comprehensive feedback :)

### Login credentials
We will discuss this. I totally would like to implement this but this leads to some kind of extra logic which is only needed for this seldom case and there only once/till the password gets changed. That's my perspective from the technical point but surely the user experience is important. I'll come back to this. Do you think a hint in the docs would be a acceptable alternative?

### Trailing slash
I encountered this while testing the application too. I'll improve this. I also tried the full API URL with the API suffix like `.../api/v2` or so. This should also get removed automatically by the application.

### Slow import
Could you please provide some spec from your system? How much CPU and RAM power does your machine have? We have general import performance topic since we are creating an huge amount of cached date - basically the whole system gets cached. This will be improved in the next steps of making the import part of Zammad more robust.

### Archived tickets
I found this article from Zendesk: https://support.zendesk.com/hc/en-us/articles/203657756-About-ticket-archiving
It states that archived tickets are accessible (as you noted) but the used API endpoint for listing tickets (which we use) does not return them 💡  Therefore we have to find a workaround for this. I'll have to dig into it and see what we can get 🔍 

### No owner set
That sounds like a bug to me. Please let me verify this. If I can reproduce it I will linke the commit to this issue. With the new testing capabilities this should be easy done.

I'm looking forward for your upcoming issues and improving the import jobs even more. ✌️

**Comment by qschneider** [comment]
Just some more minor issues
 - only the main email address of an user seems to be imported.
There seems to be no option in zammad yet to provide more than one email address per contact.

- In zendesk it is possible to add multiple contacts in cc to a ticket (and add/delete them while the ticket history progresses). This information got lost during migration und is actually a quite important feature to keep colleagues or co-workers of the customer get involved in a ticket.

Any chance to see both in the zammad in the near future (4 month) + get it imported ?

**Comment by thorsteneckel** [comment]
### Main email adress
Yes - we currently support only one identifier per user. We have a similar situation with telephone numbers which could or should also be usable to lookup/identifier users. This leads to a general thought about a multi-identifier logic which I strongly favor. However we have currently other more urgent topics to handle so there are no concrete dates or plans yet.

### CC Contacts
Do I get you right that this is some kind of a watchlist / subscription feature? Are they added to an article / comment or the ticket directly?

Sadly there is currently no active plan to implement this in the next 4 month - yet. I'll keep you updated if things are changing.

**Comment by thorsteneckel** [comment]
@qschneider  I'm trying to get grip on the archived tickets but our test instance has any archived tickets so far 👎  So I need your help to get those that are interesting for you/us. Would it be possible for you to:

1. create a file called `zendesk_debug.rb` in your Zammad folder
2. open the file and store the following content:
```ruby
Import::Zendesk::Requester.client.search(:query => "solved>#{Date.today - 120} type:ticket").all! do |ticket|
 p ticket
end
```
3. execute it via `rails r zendesk_debug.rb`

Does this list the expected tickets for you?

If possible we can also do this together via a remote session (TeamViewer or so). Just leave me a note at support@zammad.com and I will get to you.

Thanks!

**Comment by qschneider** [comment]
sure, i'ill drop you a note.

**Comment by qschneider** [comment]
I found some information about archived tickets in zendesk and the ways to access them.

https://support.zendesk.com/hc/en-us/articles/203657756-About-ticket-archiving

quote - "All ticket API endpoints except for Listing Tickets return archived tickets"

https://developer.zendesk.com/rest_api/docs/core/tickets#listing-tickets

i don't know how do you fetch the list of available tickets - at least listing tickets won't give you the archived tickets. Maybe getting the ticket-list of all agents ?

I will try to test this with postman and see what i will get.

**Comment by qschneider** [comment]
if found the solution here https://developer.zendesk.com/rest_api/docs/core/incremental_export#incremental-ticket-export

using api/v2/incremental/tickets.json?start_time=1334440800 gives me the full list of all tickets.
I just used some random date in the past (15.04.2012 = 1334440800) before the date i subscribed to zendesk.

Hope it helps, let me know when i should test this.

**Comment by qschneider** [comment]
that way you also could import tickets after the initial migration process, which would be quite nice to provide a soft migration... or just let the user decide just to start with the active tickets and import the archived ones later on.

**Comment by thorsteneckel** [comment]
Nice found! Sadly the ruby API client we use is not able to perform actions to this endpoint. I'll have to check first if we can add it with reasonable effort.

As you noted using docker please try to run the command as follows:

`zammad rails r zendesk_debug.rb`

This is needed since we are shipping rails with our OS package.

JFI: Maybe this feature is interesting for you (regarding CC functionality): https://github.com/zammad/zammad/issues/208#issuecomment-267982536

**Comment by qschneider** [comment]
thanks, `zammad run rails r zendesk_debug.rb` did it.
And yes, i got some tickets returned that were solved 120 days ago but they all shown in Zammad anyway. So still no archived tickets with the current implementation, sorry.

I will look up #208 and will let you know if this relates to my usecase.

**Comment by qschneider** [comment]
some thoughts to your questions above

- Login credentials
I think it is a good starting point from where it is right now. If it's also mentioned in the documentation, at least users have the opportunity to get the information before starting the migration. On the other hand, having it at the logging screen would make it more user friendly.

- slow import
I'am using docker (2 cores and 4GB memory) to do the testing, which runs on macOS in my case. As far i could tell during migration, cpu and network bandwidth were really low - about 15% or so. After migration finished (and indexing began?) cpu activity increased a lot more.

- main email adress
would be nice, also because otherwise data would get lost during migration.

- cc contacts
in zendesk it is applied to the ticket itself and it’s super useful to do some sort of collaboration in tickets. This is especially useful because zendesk is sending an email after each reply  - which can be super annoying, too :-) 
Also just to keep the supervisor of the a customer informed about the progress which is more a subscription feature, but still with the possibility to interact.
Or to get a colleague involved in the ticket without escalating or changing ownership.
As i’ve seen so far you are providing the cc feature when creating a new email ticket in zammad. 
You just don’t show this in the ticket or let the user add/delete contacts to it later in the process.

**Comment by qschneider** [comment]
just a thought about the api endpoints. 
If you are using the official zendesk ruby importer gem, then  this pull request https://github.com/zendesk/zendesk_api_client_rb/pull/287 might be the solution. At least this api endpoint is used somewhere in that wonderful documented piece of software :-)

**Comment by thorsteneckel** [comment]
Wow - thanks for your investigation! It gave me the idea how to solve this archive issue. You can see my changes in the last linked commit. Now it's default behavior to import all tickets - including archived 🎉  Would be great if you could check it out and test if it works for you.

Please note that I'm on vacation till 11.1.2017 and wont be able to work further on this one. I wish you a good start in 2017 and thanks a lot this far. I'll take on where I left it and come back to you.

✌️

**Comment by qschneider** [comment]
Hi, just a quick reply. Hope you also had a good start into 2017 !

I tested the import again and noticed the following.

- Import succeeded with all (active and archived) tickets. :thumbsup:
- The counter in the import dialog still shows only the active amount of tickets, f.i. XX/88 (active tickets), but there are a total of 500 tickets to import.
- Import speed is still really slow (about 45 Minutes for 500 tickets). From what i could see neither the processes inside the docker container nor the docker host where active by more than 2%.
On the zendesk-api-overview, the transfer rate is way below the max of 700 request/min.
I don't think it's client issue, at least i can't track it down on my system.

Best regards

**Comment by thorsteneckel** [comment]
Hi @qschneider - Thanks for the feedback. I'm working on the import dialog issue and let you know.

I took the time and analyzed the Zendesk import functionality:

<img width="983" alt="performancezendesk" src="https://cloud.githubusercontent.com/assets/3873515/21929295/07a64402-d990-11e6-93ad-71ec29c04275.png">

This overview shows that ~13,7% of the time (red boxes) is spend in actual import logic. The rest (~86,3%) is fetching the Zendesk API and working on the result. Which is actually pretty much. From my point of view ~66,2% is spent on HTTP requests.

In my test 144 Tickets (and other stuff) were imported in 78,93s which makes ~0,55 seconds per Ticket. Your test are ~500 Tickets in ~2700s which makes a ratio of ~0,18 seconds which is even better.

Currently I see no possibility to improve the import performance within our range 😞

**Comment by qschneider** [comment]
Hi @thorsteneckel,
Thank you for your analysis.
Btw, in my case (2770sec / 550tickets) this would be 5,4sec/ticket not 0,18, if i'am not mistaken.

If there is no way to improve performance, the max expected timeframe should be mentioned somewhere in the import dialog to inform the user, so that he can decide if he wants to invest that time right now or not. 
Maybe it also would be possible to give a rough approximation based on the actual import data?
Also it should be possible to cancel the operation at this point.

Also it would be informative to know, what would happen if someone closes the import dialog window during migration. Would this cause an interruption of migration ?

Thanks for the time you invested so far in this, i think you and the team are doing a great job!

Can't wait to switch from zendesk, which is a great service too, but they tend to move towards larger customers having larger budgets and service needs than myself.

**Comment by thorsteneckel** [comment]
🤦‍♂️  - you are right!

I really like your proposals. I'll create a new issue for this and will link it to this one since it should be available in all existing and upcoming imports.

The commit above fixes the wrong count of tickets in the import overview.

Regarding performance:
I tried to use keep-alive connections to reduce the time initializing the connection but it wasn't that successful. It reduced the time about 1 second. Since it introduced another gem dependency I discarded the code.

If you like I can provide you a customized file for measuring the performance of your import like I did?

Thanks for the feedbacks - keeps us going 🚀
